userspace/libsinsp/mpsc_priority_queue.h

```diff
@@ -90,22 +92,64 @@ class mpsc_priority_queue
 	 * element is not popped from the queue and this method returns false.
 	 */
 	template <typename Callable>
-	inline bool pop_if(const Callable& cl, OUT Elm& res)
+	inline bool try_pop_if(Elm& res, const Callable& pred)
 	{
-		elm_ptr top = m_queue_top.load();
-		if (top == nullptr || !cl(top))
+		// we check that the queue is not empty before acquiring the lock
+		if (m_queue_top == nullptr)
 		{
 			return false;
 		}
 
-		// at this point, we have a guarantee
-		// that queue.top() has priority not less than the local top,
-		// and we can pop the queue top safely
-		std::scoped_lock<Mtx> lk(m_mtx);
-		res = std::move(m_queue.top().elm);
-		m_queue.pop();
-		m_queue_top = m_queue.empty() ? nullptr : m_queue.top().elm.get();
-		return true;
+		while (true)
+		{
+			// we need to evaluate the top element against the predicate, but
+			// we must be careful in case other producers push a new element in
+			// the queue, which can potentially have more priority than the one
+			// we just checked.
+			elm_ptr top = m_queue_top.load();
+			auto should_pop = pred(*top);
+			
+			// we must not pop the element
+			if (!should_pop)
```

Comment:
side note, if there are really fast producers with old timestamps there is the risk that we wait here a little bit

Reply:
True, however it would be a matter of consistency. If there are really producers pushing higher priority events while we're trying to pop them from the main thread, then it means that we may either discard or accept the wrong event, which could be even more wrong.